README.md
Needle map to GameMaker converter
===

When designing intricate platforming for [guy games](http://delicious-fruit.com/), it's nice to use an editor like RMJ or Jtool to be able to quickly test and iterate on your designs. However, once you've finalized your design, it can be quite tedious to manually copy over every single object into GameMaker. I made this tool to automate that process.

The input file, an RMJ or Jtool map, is a text file containing a list of objects, each with an x coordinate, y coordinate, and type *number*. The output file, a GameMaker room, is an XML file where each object has an x coordinate, y coordinate, and name *string* (and other default values). The GUI lets you choose your input and output files, and also lets you map which object type numbers correspond to which object names in your GM project. It conveniently uses a file selection dialog to choose object names, so you don't have to type them out yourself.

For GameMaker 8.x projects, [Gmksplitter](https://github.com/Medo42/Gmk-Splitter) is used to decompose the binary project file to allow selection objects, and then to recompose the project file with the new room added.

[Discussion on I Wanna Community forums](http://www.iwannacommunity.com/forum/index.php?topic=1744.0)

![](http://i.imgur.com/GCyRLwe.png)


changelog.txt
--- version 2.1   2/25/2016 ---
Should now work on Windows 8 and 10 (Compiled on Windows 10, previously was on 7)
Check for new version on startup
Fix bug when importing jtool 1.2.0+ maps
Add link to tools website

--- version 2.0   11/29/2015 ---
Add GameMaker 8.x support using Gmksplit
Add jtool map support
For studio projects, output a .room.gmx file instead of modifying the .project.gmx file to import it
Redo instructions and readme files
Fix bug where template room couldn't have objects in it
GUI tweaks
Use pyinstaller to tidy up the tool's folder
Lots of code refactoring and cleanup
New name: map2gm

--- version 1.0   7/7/2015 ---
Initial release!


map2gm.py
# entry point, input sanitization, general controller stuff

import os
import requests
import sys
# map2gm modules
import gui
import convert
import localize
import util

def loc(key):
    return localize.loc(key)

def submitted(project_path, template_room_name, map_path, object_inputs):

    # input sanitization
    if project_path == '':
        return loc('project_empty')
    if template_room_name == '':
        return loc('template_room_empty')
    if map_path == '':
        return loc('map_empty')

    if not os.path.exists(project_path):
        return loc('project_nonexistent')
    project_extension = project_path.split('.')[-1]
    if project_extension == 'gmx':
        template_room_path = os.path.join(os.path.split(project_path)[0], 'rooms', template_room_name+'.room.gmx')
    else:
        template_room_path = os.path.join(util.get_application_path(), 'temp_gmksplit', 'Rooms', template_room_name+'.xml')
    if not os.path.exists(template_room_path):
        return loc('template_room_nonexistent')
    if not os.path.exists(map_path):
        return loc('map_nonexistent')

    for objectname, enabled in object_inputs.values():
        if enabled:
            if objectname == '':
                return loc('enabled_object_no_name')
            if project_extension == 'gmx':
                object_path = os.path.join(os.path.split(project_path)[0], 'objects', objectname + '.object.gmx')
                if not os.path.exists(object_path):
                    return loc('object_nonexistent') % objectname

    # build dict of object names that were enabled
    chosen_object_names = {}
    for object_name, (gm_object_name, enabled) in object_inputs.items():
        if enabled:
            chosen_object_names[object_name] = gm_object_name

    conversion_output_filename = convert.convert(project_path, template_room_path, map_path, chosen_object_names)

    # save preferences
    with open('prefs', 'w') as f:
        f.write('project|%s\n' % project_path)
        f.write('template|%s\n' % template_room_name)
        for object_name, (gm_object_name, enabled) in sorted(object_inputs.items()):
            f.write('%s|%s|%i\n' % (object_name, gm_object_name, enabled))

    if project_extension == 'gmx':
        base = loc('convert_success_gmx')
    else:
        base = loc('convert_success_gmk')
    return base % conversion_output_filename

# prompt and load language
if os.path.exists('lang'):
    with open('lang', 'r') as f:
        language = list(f)[0]
    localize.load(language)
else:
    language = gui.ask_language()
    with open('lang', 'w') as f:
        f.write(language)
    localize.load(language)
    gui.show_instructions()

# check for update
try:
    r = requests.get('http://cwpat.me/map2gm-version')
    if r.status_code == 200:
        my_version = '2.1'
        newest_version = r.json()['map2gm-version']
        if my_version != newest_version:
            gui.show_update(newest_version)
except Exception as e:
    print('Error when checking for new version.')
    print(e)

# kick off the gui
gui.run(submit_func=submitted)


convert.py
# actual conversion logic - parses map files, outputs GM files

import os
import sys
import random
import math
import subprocess
import lxml.etree as ET
# map2gm modules
import util

object_ids = {
    'block':(2,1),
    'miniblock':(None,2),
    'spikeup':(12,3),
    'spikeright':(11,4),
    'spikeleft':(10,5),
    'spikedown':(9,6),
    'miniup':(19,7),
    'miniright':(18,8),
    'minileft':(17,9),
    'minidown':(16,10),
    'save':(32,12),
    'platform':(31,13),
    'water1':(23,14),
    'water2':(30,15),
    'water3':(None,23),
    'cherry':(20,11),
    'hurtblock':(27,18),
    'vineright':(28,17),
    'vineleft':(29,16),
    'jumprefresher':(None,22),
    'bulletblocker':(None,19),
    'start':(3,20),
    'warp':(None,21),
    }

def convert(project_path, template_room_path, map_path, chosen_names):

    # build conversion dicts according to chosen object names
    rmj_to_objectname = {}
    jtool_to_objectname = {}
    for name, gm_name in chosen_names.items():
        rmj_id, jtool_id = object_ids[name]
        if rmj_id != None:
            rmj_to_objectname[str(rmj_id)] = gm_name
        jtool_to_objectname[str(jtool_id)] = gm_name

    # read instances from map file
    map_instances = []
    extension = map_path.split('.')[-1]
    if extension == 'map':
        with open(map_path) as f:
            line = f.readlines()[3]
        numbers = line[1:].split(' ')
        for i in range(0, len(numbers), 3):
            x, y, id = numbers[i:i+3]
            if id in rmj_to_objectname:
                map_instances.append((x,y,rmj_to_objectname[id]))
    elif extension == 'jmap':
        with open(map_path) as f:
            line = f.readline()
            if line[-1] == '\n':
                line = line[0:-1]
            sections = line.split('|')
        def base32string_decode(string):
            base32string = '0123456789abcdefghijklmnopqrstuv'
            result = 0
            for i, char in enumerate(string):
                charvalue = base32string.index(char)
                placevalue = math.pow(32, len(string)-i-1)
                result += charvalue * placevalue
            return str(int(result))
        for section in sections:
            if section.split(':')[0] == 'objects':
                objectstring = section.split(':')[1]
        for ypos_section in objectstring.split('-'):
            y = ypos_section[0:2]
            y = base32string_decode(y)
            y = str(int(y)-128)
            ypos_section = ypos_section[2:]
            for i in range(0, len(ypos_section), 3):
                id = ypos_section[i]
                id = base32string_decode(id)
                x = ypos_section[i+1:i+3]
                x = base32string_decode(x)
                x = str(int(x)-128)
                if id in jtool_to_objectname:
                    map_instances.append((x,y,jtool_to_objectname[id]))

    # construct valid output room name and path
    output_room_name = 'rMapImport_%s' % os.path.split(map_path)[1].split('.')[0]
    valid_chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_'
    output_room_name = ''.join([char if char in valid_chars else '_' for char in output_room_name])
    project_extension = project_path.split('.')[-1]
    if project_extension == 'gmx':
        output_room_path = os.path.join(util.get_application_path(), output_room_name+'.room.gmx')
    else:
        output_room_path = os.path.join(util.get_application_path(), 'temp_gmksplit', 'Rooms', output_room_name+'.xml')

    # create a new room file (based on template) with the instances added
    parser = ET.XMLParser(remove_blank_text=True)
    output_room_tree = ET.parse(template_room_path, parser)
    instances_element = output_room_tree.getroot().find('instances')
    for x, y, name in map_instances:
        if project_extension == 'gmx':
            attrib = {}
            attrib['x'] = x
            attrib['y'] = y
            attrib['objName'] = name
            attrib['locked'] = '0'
            attrib['code'] = ''
            attrib['scaleX'] = '1'
            attrib['scaleY'] = '1'
            attrib['colour'] = '4294967295'
            attrib['rotation'] = '0'
            attrib['name'] = 'inst_'+''.join([random.choice('0123456789ABCDEF') for i in range(8)])
            ET.SubElement(instances_element, 'instance', attrib=attrib)
        else:
            inst_elt = ET.Element('instance')
            obj_elt = ET.SubElement(inst_elt, 'object')
            obj_elt.text = name
            ET.SubElement(inst_elt, 'position', x=x, y=y)
            ET.SubElement(inst_elt, 'creationCode')
            locked_elt = ET.SubElement(inst_elt, 'locked')
            locked_elt.text = 'false'
            instances_element.append(inst_elt)
    output_room_tree.write(output_room_path, pretty_print = True)

    # if gmksplit project, add room to room list and recompose project file
    if project_extension != 'gmx':
        room_resources_path = os.path.join(util.get_application_path(), 'temp_gmksplit', 'Rooms', '_resources.list.xml')
        parser = ET.XMLParser(remove_blank_text=True)
        room_resources_tree = ET.parse(room_resources_path, parser)
        room_already_in_resource_list = False
        for room_element in room_resources_tree.getroot():
            if room_element.attrib['name'] == output_room_name:
                room_already_in_resource_list = True
        if not room_already_in_resource_list:
            ET.SubElement(room_resources_tree.getroot(), 'resource', name=output_room_name, type='RESOURCE')
            room_resources_tree.write(room_resources_path, pretty_print = True)

        output_project_name = os.path.split(project_path)[1].split('.')[-2]
        output_project_name += '_' + ''.join([random.choice('abcdefghijklmnopqrstuvwxyz') for i in range(5)]) + '.' + project_extension
        output_project_path = os.path.join(util.get_application_path(), output_project_name)
        subprocess.call(os.path.join(util.get_application_path(), 'gmksplitter\\gmksplit.exe temp_gmksplit "%s"' % output_project_path))

    if project_extension == 'gmx':
        return output_room_name
    else:
        return output_project_name


rmj2gms.py
# entry point, input sanitization, general controller stuff

import os
import gui, convert, localize

def loc(key):
    return localize.loc(key)

# object_inputs is a map: {object_name: (gm_object_name, enabled)}
def submitted(project_path, template_room_name, map_path, object_inputs):

    # input sanitization
    if project_path == '' or template_room_name == '' or map_path == '':
        return loc('error_top_field_empty')
    if not os.path.exists(project_path):
        # TODO localize
        return 'project does not exist'
    if not os.path.exists(map_path):
        # TODO localize
        return 'map does not exist'
    template_room_path = os.path.join(os.path.split(project_path)[0], 'rooms', template_room_name+'.room.gmx')
    if not os.path.exists(template_room_path):
        # TODO localize
        return 'template room does not exist'
    for objectname, enabled in object_inputs.values():
        if enabled:
            if objectname == '':
                return loc('error_object_no_name')
            fn = os.path.join(os.path.split(project_path)[0], 'objects', objectname + '.object.gmx')
            if not os.path.exists(fn):
                return loc('error_nonexistent_object') % objectname

    # build dict of object names that were enabled
    chosen_object_names = {}
    for object_name, (gm_object_name, enabled) in object_inputs.items():
        if enabled:
            chosen_object_names[object_name] = gm_object_name

    output_room_name = convert.convert(project_path, template_room_path, map_path, chosen_object_names)

    # save preferences
    with open('prefs', 'w') as f:
        f.write('project|%s\n' % project_path)
        f.write('template|%s\n' % template_room_name)
        for object_name, (gm_object_name, enabled) in sorted(object_inputs.items()):
            f.write('%s|%s|%i\n' % (object_name, gm_object_name, enabled))

    return loc('convert_successful') + ' ' + output_room_name

# prompt and load language
if os.path.exists('lang'):
    with open('lang', 'r') as f:
        language = list(f)[0]
    localize.load(language)
else:
    language = gui.ask_language()
    with open('lang', 'w') as f:
        f.write(language)
    localize.load(language)
    gui.show_instructions()

# kick off the gui
gui.run(submit_func=submitted)


gui.py
# controls window for user input

import sys
import os
import subprocess
import shutil
import traceback
import webbrowser
import tkinter as tk
from tkinter import filedialog, messagebox
# map2gm modules
import localize
import util

def loc(key):
    return localize.loc(key)

def delete_temp_gmksplit_if_necessary():
    temp_gmksplit_path = os.path.join(util.get_application_path(), 'temp_gmksplit')
    if os.path.exists(temp_gmksplit_path):
        shutil.rmtree(temp_gmksplit_path)

def gmksplit_decompose(project_path):
    delete_temp_gmksplit_if_necessary()
    subprocess.call(os.path.join(util.get_application_path(), 'gmksplitter\\gmksplit.exe "%s" temp_gmksplit' % project_path))

def call_func_loading_popup(func, text):
    popup = tk.Toplevel()
    popup.wm_title('')
    popup.resizable(False, False)
    icon_image = tk.Image('photo', file='images/icon.png')
    popup.tk.call('wm','iconphoto',popup._w,icon_image)
    label = tk.Label(popup, text=text)
    label.grid(padx=30, pady=30)

    x = root.winfo_x() + 50
    y = root.winfo_y() + 50
    popup.geometry('+%d+%d' % (x, y))
    popup.update()
    func()
    popup.destroy()

def update_object_widgets(check_enabled, entry, button):
    if check_enabled:
        state = tk.NORMAL
    else:
        state = tk.DISABLED
    entry.configure(state=state)
    button.configure(state=state)

def ask_path(entry, dialogtitle, filetypes, initialdir, format_func):
    path = filedialog.askopenfilename(filetypes=filetypes,title=dialogtitle,initialdir=initialdir)
    if path != '':
        path = format_func(path)
        entry.delete(0, tk.END)
        entry.insert(0, path)
        entry.xview(tk.END)

def project_row_clicked(entry, project_entry):
    dialogtitle = loc('open_project')
    filetypes = [('GameMaker Studio, 8.1, or 8.0 project', '*.gm*')]
    initialdir = util.get_application_path()
    format_func = lambda path: path

    lastpath = entry.get()
    ask_path(entry, dialogtitle, filetypes, initialdir, format_func)

    filetype = entry.get().split('.')[-1]
    if filetype != 'gmx' and entry.get() != lastpath and os.path.exists(entry.get()):
        call_func_loading_popup(lambda: gmksplit_decompose(entry.get()), loc('gmksplit_working'))

def template_row_clicked(entry, project_entry):
    filetype = project_entry.get().split('.')[-1]
    if filetype == 'gmx':
        filetypes = [('GameMaker: Studio room', '*.room.gmx')]
        initialdir = os.path.join(os.path.split(project_entry.get())[0],'rooms')
    else:
        filetypes = [('Gmksplit room', '*.xml')]
        initialdir = os.path.join(util.get_application_path(),'temp_gmksplit','Rooms')
        if not os.path.exists(initialdir):
            call_func_loading_popup(lambda: gmksplit_decompose(project_entry.get()), loc('gmksplit_working'))
    dialogtitle = loc('open_template_room')
    format_func = lambda path: os.path.split(path)[1].split('.')[0]
    ask_path(entry, dialogtitle, filetypes, initialdir, format_func)

def map_row_clicked(entry, project_entry):
    dialogtitle = loc('open_map')
    filetypes = [('Jtool or Record My Jumps map', '*.*map')]
    initialdir = util.get_application_path()
    format_func = lambda path: path
    ask_path(entry, dialogtitle, filetypes, initialdir, format_func)

def object_row_clicked(entry, project_entry):
    filetype = project_entry.get().split('.')[-1]
    if filetype == 'gmx':
        filetypes = [('GameMaker: Studio object', '*.object.gmx')]
        initialdir = os.path.join(os.path.split(project_entry.get())[0],'objects')
    else:
        filetypes = [('Gmksplit object', '*.xml')]
        initialdir = os.path.join(util.get_application_path(),'temp_gmksplit','Objects')
        if not os.path.exists(initialdir):
            call_func_loading_popup(lambda: gmksplit_decompose(project_entry.get()), loc('gmksplit_working'))
    dialogtitle = loc('open_object')
    format_func = lambda path: os.path.split(path)[1].split('.')[0]
    ask_path(entry, dialogtitle, filetypes, initialdir, format_func)

def entry_button_row(row, labeltext):
    label = tk.Label(root,text=labeltext)
    entry = tk.Entry(root)
    button = tk.Button(root,image=folder_image,width=35,height=25)
    label.grid(row=row,column=0,sticky=tk.E)
    entry.grid(row=row,column=1,sticky=tk.EW)
    button.grid(row=row,column=2,sticky=tk.W)
    return (entry, button)

def show_update(newversion):
    window = tk.Tk()
    window.resizable(False, False)
    window.geometry('350x150')
    window.wm_title('')
    label = tk.Label(window,text=loc('update_label').format(version=newversion))
    label.grid(row=0,column=0,sticky=tk.NSEW,padx=20,pady=20)
    def callback():
        webbrowser.open('https://www.bit.ly/needle-map-to-gm')
        window.destroy()
    button = tk.Button(window, text=loc('update_button'), command=callback)
    button.grid(row=1,column=0,sticky=tk.NSEW,padx=20,pady=20)
    window.mainloop()

def ask_language():
    global chosen
    window = tk.Tk()
    window.resizable(False, False)
    window.geometry('300x100')
    window.wm_title('')
    window.protocol('WM_DELETE_WINDOW', lambda: None)
    icon_image = tk.Image('photo', file='images/icon.png')
    window.tk.call('wm','iconphoto',window._w,icon_image)
    def choose(c):
        global chosen
        window.destroy()
        chosen = c
    tk.Button(window, text='English', command=lambda: choose('English')).grid(row=0,column=0,sticky=tk.NSEW,padx=10,pady=10)
    tk.Button(window, text='日本語', command=lambda: choose('Japanese')).grid(row=0,column=1,sticky=tk.NSEW,padx=10,pady=10)
    window.grid_columnconfigure(0, weight=1)
    window.grid_columnconfigure(1, weight=1)
    window.grid_rowconfigure(0, weight=1)
    window.mainloop()
    return chosen

def change_language(language):
    with open('lang', 'w') as f:
        f.write(language)
    os.execl(sys.executable, sys.executable, * sys.argv)

def show_instructions():
    os.startfile('instructions.html')

def show_readme():
    os.startfile('readme.txt')

def submit(submit_function, convert_button, *args):
    convert_button.config(text=loc('button_convert_working') + '  ')
    root.update()

    filetype = args[0].split('.')[-1]
    if filetype != 'gmx' and not os.path.exists(os.path.join(util.get_application_path(),'temp_gmksplit')):
        gmksplit_decompose(args[0])

    try:
        message = submit_function(*args)
    except:
        message = loc('python_error') + '\n' + traceback.format_exc()
        traceback.print_exc()

    convert_button.config(text=loc('button_convert') + '  ')
    root.update()
    messagebox.showinfo('', message)

def run(submit_func):
    # init stuff
    global root, folder_image
    root = tk.Tk()
    folder_image = tk.Image('photo', file='images/folder.png')

    # top three input fields
    current_row = 0
    labeltext = loc('label_map')
    map_textbox, map_button = entry_button_row(current_row, labeltext)

    current_row += 1
    labeltext = loc('label_project')
    project_textbox, project_button = entry_button_row(current_row, labeltext)

    current_row += 1
    labeltext = loc('label_template_room')
    template_textbox, template_button = entry_button_row(current_row, labeltext)

    map_button.config(command=lambda: map_row_clicked(map_textbox, project_textbox))
    template_button.config(command=lambda: template_row_clicked(template_textbox, project_textbox))
    project_button.config(command=lambda: project_row_clicked(project_textbox, project_textbox))

    # object input fields
    current_row += 1
    object_names = [
        'block','spikeup','spikeright','spikeleft','spikedown',
        'miniblock','miniup','miniright','minileft','minidown',
        'save','platform','water1','water2','water3','cherry','hurtblock',
        'vineright','vineleft','jumprefresher','bulletblocker','start','warp',
        ]
    object_images = [(o,'images/%s.png'%o) for o in object_names]
    objectrowheight = 40
    frameheight = 5 * objectrowheight
    canvasheight = (len(object_images) - 1) * objectrowheight
    frame = tk.Frame(root,height=frameheight)
    frame.grid(row=current_row,column=0,columnspan=3,padx=5,pady=5,sticky=tk.NS)
    canvas = tk.Canvas(frame,scrollregion=(0,0,0,canvasheight),height=frameheight, width=340, yscrollincrement=objectrowheight)
    vbar=tk.Scrollbar(frame,orient=tk.VERTICAL)
    vbar.pack(side=tk.RIGHT,fill=tk.Y)
    vbar.config(command=canvas.yview)
    canvas.config(yscrollcommand=vbar.set)
    canvas.pack(side=tk.LEFT,expand=True,fill=tk.BOTH,pady=20)

    objectrow = 0
    object_widgets = {}
    for objectname, imagepath in object_images:
        photo = tk.PhotoImage(file=imagepath)
        w = tk.Label(root,image=photo)
        w.photo = photo # to prevent it from being garbage collected
        canvas.create_window((25,0+objectrow*objectrowheight),anchor=tk.CENTER,window=w)

        e = tk.Entry(root,state=tk.DISABLED)
        canvas.create_window((60,0+objectrow*objectrowheight),anchor=tk.W,window=e,width=120)

        cmd = lambda entry=e: object_row_clicked(entry, project_textbox)
        b = tk.Button(root,image=folder_image,command=cmd,state=tk.DISABLED)
        canvas.create_window((190,0+objectrow*objectrowheight),anchor=tk.W,window=b,width=40,height=30)

        v = tk.BooleanVar()
        v.set(False)
        cmd = lambda var=v,entry=e,button=b: update_object_widgets(var.get(),entry,button)
        c = tk.Checkbutton(root,text=loc('label_object_enabled'),variable=v,command=cmd)
        canvas.create_window((240,0+objectrow*objectrowheight),anchor=tk.W,window=c)

        object_widgets[objectname] = (e, v, b)
        objectrow += 1

    # convert button
    current_row += 1
    filler_label = tk.Label(root)
    filler_label.grid(row=current_row,column=0,columnspan=2,sticky=tk.NSEW)
    icon_image = tk.Image('photo', file='images/icon.png')
    convert_button = tk.Button(root,text=loc('button_convert') + '  ',image=icon_image,compound=tk.RIGHT,width=150)
    cmd = lambda: submit(submit_func, convert_button, project_textbox.get(), template_textbox.get(), map_textbox.get(), {k: (v[0].get(),v[1].get()) for (k,v) in object_widgets.items()})
    convert_button.configure(command=cmd)
    convert_button.grid(row=current_row,column=1,columnspan=2,sticky=tk.SE)

    # load values from prefs file
    if os.path.exists('prefs'):
        with open('prefs', 'r') as f:
            for line in f:
                args = line[:-1].split('|')
                args += [''] * (3 - len(args))
                type, value, checked = args
                if type == 'template':
                    template_textbox.insert(0, value)
                    template_textbox.xview(tk.END)
                elif type == 'project':
                    project_textbox.insert(0, value)
                    project_textbox.xview(tk.END)
                else:
                    entry, var, button = object_widgets[type]
                    entry.configure(state=tk.NORMAL)
                    entry.insert(0, value)
                    var.set(int(checked))
                    update_object_widgets(var.get(), entry, button)

    delete_temp_gmksplit_if_necessary()

    # menu bar
    menubar = tk.Menu(root)
    optionsmenu = tk.Menu(menubar, tearoff=False)
    menubar.add_cascade(label=loc('menu_toplevel'), menu=optionsmenu)
    optionsmenu.add_command(label=loc('menu_instructions'), command=show_instructions)
    optionsmenu.add_command(label=loc('menu_readme'), command=show_readme)
    optionsmenu.add_command(label=loc('menu_forum_thread'), command=lambda: webbrowser.open('https://www.bit.ly/needle-map-to-gm'))
    root.config(menu=menubar)

    # configure window and enter its main loop
    root.update()
    root.minsize(root.winfo_width(), root.winfo_height())
    root.grid_columnconfigure(0, weight=1, minsize=110)
    root.grid_columnconfigure(1, weight=8, minsize=150)
    root.grid_columnconfigure(2, weight=1, minsize=35)
    root.grid_rowconfigure(current_row-1, weight=1)
    root.resizable(True, True)
    root.wm_title(loc('title'))
    root.tk.call('wm','iconphoto',root._w,icon_image)
    root.mainloop()
